README.md
[<img src="https://cloud.githubusercontent.com/assets/16360374/23640973/ac0fe0a0-02a5-11e7-91a1-fc9eb72514de.png" height="120" title="PlayMusic"/>](https://github.com/tterb/PlayMusic)
----------------------------------------------------------------------------

<!-- Badges -->
<p align="right">
  <a href="https://badge.fury.io/gh/tterb%2FPlayMusic">
	  <img src="https://badge.fury.io/gh/tterb%2FPlayMusic.svg" height="21" alt="version">
  </a>
  <a href="http://www.gnu.org/licenses/agpl-3.0">
		<img src="https://img.shields.io/badge/license-AGPL-blue.svg" height="21" title="License">
  </a>
  <a href="https://github.com/tterb/PlayMusic/issues">
	  <img src="https://img.shields.io/github/issues-raw/tterb/PlayMusic.svg?maxAge=25000" height="21" title="Issues">
  </a>
  <a href="http://www.deviantart.com/art/PlayMusic-620128929">
  	<img src="https://img.shields.io/badge/downloads-%207%2C054-green.svg" height="21" title="Downloads">
  </a>
  <span class="badge-paypal">
    <a href="https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=E6RKPR34SH6CU" title="Donate to this project using Paypal">
          <img src="https://img.shields.io/badge/paypal-donate-yellow.svg" height="21" title="Donate">
    </a>
  </span>
</p>


<!-- Screenshot -->
<p align="center">
  <img src="https://cloud.githubusercontent.com/assets/16360374/22547134/1a4e0da0-e8f5-11e6-9802-08228a312ec6.gif" width="810">
</p>


<!-- ------------------------------------------------------------------------------ -->
<br>

[<img title="" alt="Description" src="https://cloud.githubusercontent.com/assets/16360374/22623856/4459294c-eb1e-11e6-86e0-7a8392da4061.png" width="900" />](#description)

  This [Rainmeter] skin provides [Google Play Music] users the ability to access their active media tastefully on their desktop. Featuring three variants, this skin displays the Album Artwork accompanied by the Title, Artist, and Album of the song being played, as well as the status of both the *shuffle* and *repeat* features.
  Additionally, this skin also includes a variety of supplementary features including a customizable visualizer, scalability, as well as volume control and minimization functionalities.

#### In Progress:
  * Add requested support for **[Spotify]** as well as other media players
  * Improve accent color adaptability for all album artwork
  * Add performance settings and further customization to visualizer
<br><br>

<!-- <p align="center">
  <a href="https://github.com/tterb/PlayMusic/releases/latest">
    <img src="https://cloud.githubusercontent.com/assets/16360374/25461398/9aa72f6a-2a9d-11e7-948d-5c8691b2392c.png" width= "200" alt="Download PlayMusic"/>
  </a>
<p>   -->

<!-- ***Note:***&nbsp; Due to current limitations, it's not possible to integrate remote control of GPMDP into the skin, though the developers have been made aware of this feature request and I believe this functionality can be expected in future updates.
<br> -->

------------------------------------------------------------------------------

[<img title="" alt="Features" src="https://cloud.githubusercontent.com/assets/16360374/22623859/446dbe70-eb1e-11e6-8f95-ec36076d7f52.png" width="900" />](#features)
<p align="center">
	<img src="https://cloud.githubusercontent.com/assets/16360374/23335412/21511926-fb6a-11e6-8bb5-249567ecd85a.png" height="240" title="Variants">
</p>
<!-- ### Variants   -->
<p align="center">
<img src="https://cloud.githubusercontent.com/assets/16360374/23576129/3d750194-0051-11e7-9b00-a06a617dda82.png" height="42" title="Variants"/>
</p>
<p align="center">
  PlayMusic features square, round, and landscape variants, which in conjunction with its <br>  added scalability settings, guarantee its ability to compliment any desktop layout.
</p><br>

<!-- ### Settings Menu -->
<p align="center">
<img src="https://cloud.githubusercontent.com/assets/16360374/22628809/c039003e-eb8f-11e6-91f3-34bb8af4364a.png" height="42" title="Settings Menu"/>
</p>
<p align="center">
  In order to facilitate easy customization, PlayMusic includes a highly intuitive settings menu, <br>  making configuration and modification a breeze. This menu will be launched upon installation, <br>  though it can also be accessed at any time by pressing the &nbsp;<img src="https://cloud.githubusercontent.com/assets/16360374/21559428/fe6a3654-ce00-11e6-8d2c-ba8570e3132c.png" height="12">&nbsp; on any of the skin variants.
</p><br>

<!-- ### Accent Colors -->
<p align="center">
<img src="https://cloud.githubusercontent.com/assets/16360374/23576126/3d745d84-0051-11e7-864a-2e0f82ece29e.png" height="42" title="Accent Colors"/>
</p>
<p align="center">
The addition of selectable accent colors allows users to add a splash of color to their skin, <br>  while also allowing PlayMusic to tastefully adapt to any theme, layout, or wallpaper.
</p><br>

<!-- ### Visualizer -->
<p align="center">
<img src="https://cloud.githubusercontent.com/assets/16360374/23576127/3d74bffe-0051-11e7-8638-e0df51683824.png" height="42" title="Visualizer"/>
</p>
<p align="center">
  In addition to providing added functionality to the skin, the recent adoption of a highly <br>  customizable visualizer also allows one more way of making this skin your own.
</p><br>

<!-- ### Minimization   -->
<!-- <p align="center">
<img src="https://cloud.githubusercontent.com/assets/16360374/23576130/3d78fd44-0051-11e7-816a-0cca182afa22.png" height="42" title="Minimization"/>
</p>
<p align="center">
  By default, the skin displays a <img src="https://cloud.githubusercontent.com/assets/16360374/22850002/fb77e47c-efb7-11e6-9199-b6b027504656.png" height="23"/>  icon whenever GPMDP is inactive, conserving useful <br>  desktop space, while also serving as a convenient launcher for the application. Though this <br>  functionality, as well as the displayed icon can be modified or disabled in the settings menu.
</p><br> -->

<!-- ### Volume control -->
<p align="center">
<img src="https://cloud.githubusercontent.com/assets/16360374/23576125/3d72d64e-0051-11e7-8b4f-114b5b3103ee.png" height="42" title="Volume Control"/>
</p><p align="center">
  This skin also features the ability to easily adjust your computers volume by simply <br>  scrolling up or down on the mouse-wheel with your cursor over the skin.
</p>

<br>

------------------------------------------------------------------------------

[<img title="" alt="Setup" src="https://cloud.githubusercontent.com/assets/16360374/22623857/445972e4-eb1e-11e6-9608-fce5c7d4f52d.png" width="900" />](#setup)

  This skin requires you have **[Google Play Music Desktop Player]** installed to function properly.
  After installation, you must also enable **Playback API** in GPMDP's *Desktop Settings*, shown in the image below, as this allows GPMDP to interface with external applications.

  <p align="center">
    <img src="https://cloud.githubusercontent.com/assets/16360374/22547192/66e31a98-e8f5-11e6-82d5-20c956411f50.png" width="720" title="Desktop Settings">
  </p>

------------------------------------------------------------------------------

[<img title="" alt="Documentation" src="https://cloud.githubusercontent.com/assets/16360374/22623858/446b33bc-eb1e-11e6-96ae-2494c5072887.png" width="900" />](#documentation)

  * [License](docs/LICENSE.md)
  * [Changelog](docs/CHANGELOG.md)
  * [Contributing](docs/CONTRIBUTING.md)
  * [Credits](docs/CREDITS.md)

<br>

------------------------------------------------------------------------------
<p align="right">
  <a href="https://gitter.im/tterb/PlayMusic?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge">
      <img src="https://badges.gitter.im/tterb/PlayMusic.svg" height="21" title="Gitter">
  </a>
</p>


<!---------------------------------- Links ------------------------------------>

[Rainmeter]: https://www.rainmeter.net/
[Google Play Music]: https://play.google.com/music
[Google Play Music Desktop Player]: http://www.googleplaymusicdesktopplayer.com/
[Spotify]: https://www.spotify.com/


docs/CHANGELOG.md
# Change Log
<br>

## Version 1.3.0

  * Added ***Spotify*** integrations to PlayMusic
  * Improved settings-menu state preservation and fluidity
  * Added [GPMDP-plugin](https://github.com/tjhrulz/GPMDP-Plugin) integration to enable media controls
  * Optimized skin update processes and performance
  * Added custom icon for Spotify
  * Modified skin context-menu to adapt to active media source
  * Added adaptive accent color functionality
  * Improved skin update processes to resource utilization

------------------------------------------------------------------------------

## Version 1.2.1

  * Fix major issues with *.rmskin* package

------------------------------------------------------------------------------

## Version 1.2.0

  * Added **Round** skin variant
  * Added a configurable visualizer to skin
  * Improved icons for  minimized state
  * Added volume control functionality
  * Improved accent color implementation
  * Added modular scaling to all skin variants
  * Overhauled the settings menu to accommodate further customization
  * Added configurable options to improve text visibility on bright or elaborate album covers

------------------------------------------------------------------------------


## Version 1.0.6

  * Added selectable accent colors
  * Improved Settings menu styling
  * Added configurable minimized icon styles
  * Improved text readability on light album covers (square)
  * Modified status icons for improved visibility
  * Increased *Artist* text size (square)
  * Fixed off-center placement of the time bar
  * Aligned skin variant placements
  * Fixed issue #6 - album artwork now conforms to skin parameters

------------------------------------------------------------------------------


assets/js/main.js
/*
	Forty by HTML5 UP
	html5up.net | @ajlkn
	Free for personal and commercial use under the CCA 3.0 license (html5up.net/license)
*/

(function($) {

    skel.breakpoints({
        xlarge: '(max-width: 1680px)',
        large: '(max-width: 1280px)',
        medium: '(max-width: 980px)',
        small: '(max-width: 736px)',
        xsmall: '(max-width: 480px)',
        xxsmall: '(max-width: 360px)'
    });

    /**
     * Applies parallax scrolling to an element's background image.
     * @return {jQuery} jQuery object.
     */
    $.fn._parallax = (skel.vars.browser == 'ie' || skel.vars.browser == 'edge' || skel.vars.mobile) ? function() {
        return $(this)
    } : function(intensity) {

        var $window = $(window),
            $this = $(this);

        if (this.length == 0 || intensity === 0)
            return $this;

        if (this.length > 1) {

            for (var i = 0; i < this.length; i++)
                $(this[i])._parallax(intensity);

            return $this;

        }

        if (!intensity)
            intensity = 0.25;

        $this.each(function() {

            var $t = $(this),
                on, off;

            on = function() {

                $t.css('background-position', 'center 100%, center 100%, center 0px');

                $window
                    .on('scroll._parallax', function() {
                        var pos = parseInt($window.scrollTop()) - parseInt($t.position().top);

                        $t.css('background-position', 'center ' + (pos * (-1 * intensity)) + 'px');
                    });

            };

            off = function() {
                $t
                    .css('background-position', '');

                $window
                    .off('scroll._parallax');
            };

            skel.on('change', function() {

                if (skel.breakpoint('medium').active)
                    (off)();
                else
                    (on)();

            });

        });

        $window
            .off('load._parallax resize._parallax')
            .on('load._parallax resize._parallax', function() {
                $window.trigger('scroll');
            });

        return $(this);

    };

    $(function() {

        var $window = $(window),
            $body = $('body'),
            $wrapper = $('#wrapper'),
            $header = $('#header'),
            $banner = $('#banner');

        // Disable animations/transitions until the page has loaded.
        $body.addClass('is-loading');

        $window.on('load pageshow', function() {
            window.setTimeout(function() {
                $body.removeClass('is-loading');
            }, 100);
        });

        // Clear transitioning state on unload/hide.
        $window.on('unload pagehide', function() {
            window.setTimeout(function() {
                $('.is-transitioning').removeClass('is-transitioning');
            }, 250);
        });

        // Fix: Enable IE-only tweaks.
        if (skel.vars.browser == 'ie' || skel.vars.browser == 'edge')
            $body.addClass('is-ie');

        // Fix: Placeholder polyfill.
        $('form').placeholder();

        // Prioritize "important" elements on medium.
        skel.on('+medium -medium', function() {
            $.prioritize(
                '.important\\28 medium\\29',
                skel.breakpoint('medium').active
            );
        });

        // Scrolly.
        $('.scrolly').scrolly({
            offset: function() {
                return $header.height() - 2;
            }
        });

        // Tiles.
        var $tiles = $('.tiles > article');

        $tiles.each(function() {

            var $this = $(this),
                $image = $this.find('.image'),
                $img = $image.find('img'),
                $link = $this.find('.link'),
                x;

            // Image.

            // Set image.
            $this.css('background-image', 'url(' + $img.attr('src') + ')');

            // Set position.
            if (x = $img.data('position'))
                $image.css('background-position', x);

            // Hide original.
            $image.hide();

            // Link.
            if ($link.length > 0) {

                $x = $link.clone()
                    .text('')
                    .addClass('primary')
                    .appendTo($this);

                $link = $link.add($x);

                $link.on('click', function(event) {

                    var href = $link.attr('href');

                    // Prevent default.
                    event.stopPropagation();
                    event.preventDefault();

                    // Start transitioning.
                    $this.addClass('is-transitioning');
                    $wrapper.addClass('is-transitioning');

                    // Redirect.
                    window.setTimeout(function() {

                        if ($link.attr('target') == '_blank')
                            window.open(href);
                        else
                            location.href = href;

                    }, 500);

                });
            }
        });

        // Header.
        if (skel.vars.IEVersion < 9)
            $header.removeClass('alt');

        if ($banner.length > 0 &&
            $header.hasClass('alt')) {

            $window.on('resize', function() {
                $window.trigger('scroll');
            });

            $window.on('load', function() {

                $banner.scrollex({
                    bottom: $header.height() + 10,
                    terminate: function() {
                        $header.removeClass('alt');
                    },
                    enter: function() {
                        $header.addClass('alt');
                    },
                    leave: function() {
                        $header.removeClass('alt');
                        $header.addClass('reveal');
                    }
                });

                window.setTimeout(function() {
                    $window.triggerHandler('scroll');
                }, 100);

            });

        }

        // Banner.
        $banner.each(function() {

            var $this = $(this),
                $image = $this.find('.image'),
                $img = $image.find('img');

            // Parallax.
            $this._parallax(0.275);

            // Image.
            if ($image.length > 0) {

                // Set image.
                $this.css('background-image', 'url(' + $img.attr('src') + ')');

                // Hide original.
                $image.hide();
            }

        });

        // Menu.
        var $menu = $('#menu'),
            $menuInner;

        $menu.wrapInner('<div class="inner"></div>');
        $menuInner = $menu.children('.inner');
        $menu._locked = false;

        $menu._lock = function() {

            if ($menu._locked)
                return false;

            $menu._locked = true;

            window.setTimeout(function() {
                $menu._locked = false;
            }, 350);

            return true;

        };

        $menu._show = function() {

            if ($menu._lock())
                $body.addClass('is-menu-visible');

        };

        $menu._hide = function() {

            if ($menu._lock())
                $body.removeClass('is-menu-visible');

        };

        $menu._toggle = function() {

            if ($menu._lock())
                $body.toggleClass('is-menu-visible');

        };

        $menuInner
            .on('click', function(event) {
                event.stopPropagation();
            })
            .on('click', 'a', function(event) {

                var href = $(this).attr('href');

                event.preventDefault();
                event.stopPropagation();

                // Hide.
                $menu._hide();

                // Redirect.
                window.setTimeout(function() {
                    window.location.href = href;
                }, 250);

            });

        $menu
            .appendTo($body)
            .on('click', function(event) {

                event.stopPropagation();
                event.preventDefault();

                $body.removeClass('is-menu-visible');

            })
            .append('<a class="close" href="#menu">Close</a>');

        $body
            .on('click', 'a[href="#menu"]', function(event) {

                event.stopPropagation();
                event.preventDefault();

                // Toggle.
                $menu._toggle();

            })
            .on('click', function(event) {

                // Hide.
                $menu._hide();

            })
            .on('keydown', function(event) {

                // Hide on escape.
                if (event.keyCode == 27)
                    $menu._hide();

            });

    });
    $(function() {
        var fonts = 'https://fonts.googleapis.com/css?family=Source+Sans+Pro:300,300italic,400,400italic,600,600italic,700,700italic,900,900italic|Open+Sans:300,300italic,400,400italic,600,600italic,700,700italic,800,800italic|Source+Code+Pro:200,200italic,300,300italic,600,600italic,700,700italic'
        if (document.createStyleSheet) document.createStyleSheet(fonts);
        else $("head").append($("<link rel='stylesheet' href='" + fonts + "' type='text/css' media='screen' />"));
    });

})(jQuery);
